<head>
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <link href="fold.css" rel="stylesheet">
</head>
<body>
    <h1>
        USM<sub style='font-size:large'>fold</sub>
        <sup style="font-size:large;font-weight:normal">
            [<a href="https://github.com/usm/fold" target="_blank">code</a>
            ] [<a href="https://observablehq.com/@episphere/usm" target="_blank">notebook</a>
            ] [<a href="https://github.com/usm/fold/issues" target="_blank">issues</a>
            ]
        
        
        
        </sup>
    </h1>
    <p>USM-fold: Bidirectional Fractal Position Encoding of Biological Sequences.</p>
    <hr>
    <h2>
        A. Universal Sequence Map (USM) generalization<br>of Chaos Game Representation (CGR)
    </h2>
    <p>
        Let's start by loading the USM module, <code>Umod = (await import(`https://usm.github.io/3/usm.mjs`))</code>
        . Give it a try in the <a href="https://developer.chrome.com/docs/devtools/console" target="_blank">browser console</a>
        ). Alternatively, <code>Umod</code>
        , can be engaged graphically to instantiate a small forward USM reproducing the original Chaos Game Representation procedure (CGR). Let's start with a short nucleotide sequence, say "GATTACA", and then add or remove nucleotides anyway you see fit.
    
    
    
    </p>
    <div id="divParms">
        <!--<b> Direction:</b> <input type="radio" id="direction_forward" name="direction" value="forward" checked="true"> forward 
            <input type="radio" id="direction_backward" name="direction"value="backward"> backward;-->
        <b>Seed:</b>
        <input type="radio" id="seed_middle" name="seed" value="middle" checked="true">
        middle (CGR)
            <input type="radio" id="seed_circular" name="seed" value="circular">
        circular
            <input type="radio" id="seed_bidirectional" name="seed" value="bidirectional">
        bidirectional ;
            <b>Size: </b>
        <input id="plotSize" type="number" min=0 max=1024 value=256>
        <input type="range" id="plotRange" min=0 max=1024 value=256>
        <textarea id="taCGR" style="width:100%;font-size:large">GATTACA</textarea>
    </div>
    <!--<div id="divPlotUSM"></div>-->
    <table>
        <tr>
            <td id="forwardACGT"></td>
            <td id="backwardACGT"></td>
        </tr>
    </table>
    <!--
    <span id="forwardACGT"></span>
    <span id="backwardACGT"></span>
    -->
    <div id="divFold">
        <hr>
        <h2>B. Density distribution</h2>
        <p>
            Each iterated direction, forward and backward, populate a unit hypercube (a unit square in the case of DNA). For the specific case of CGR (forward USM with [1/2,1/2] seed) this is designated as a frequency CGR (FCGR). In the frequency mapping below the bidirectional seeding procedure is used. 
            The default sequence is <a href="https://www.ncbi.nlm.nih.gov/sviewer/viewer.fcgi?id=399923581&db=nuccore&report=fasta&extrafeat=null&conwithfeat=on&hide-cdd=on&retmode=html&withmarkup=on&tool=portal&log$=seqview" target="_blank">NG_007726.3</a>
            for Homo sapiens epidermal growth factor receptor (EGFR), RefSeqGene (<a href="https://www.ncbi.nlm.nih.gov/nuccore/399923581" target="_blank">LRG_304</a>
            ) on chromosome 7.
            To try with your own sequences, paste them in the text area and click the button "Density".
        </p>
        <button id="densityButton" style="color:gray">Density</button>
        <b>Quadrants</b>
        : <input id="numQuadrants" type="number" value=256 style="color:blue;width:60" min=2 max=1024>
        <input id="rangeQuadrants" type="range" value=256 min=2 max=1024 style="width:200">
        , <b>Color</b>
        : <input id="colorCheckbox" type="checkbox" checked=true>
        ; <b>resize</b>
        : <input id="resizeCheckbox" type="checkbox" checked=true>
        <br>
        <b>Sequence length</b>
        : <span id="seqLength">...</span>
        , <b>Bins</b>
        : <span id="numBins">65536</span>
        , <b>n-gram length</b>
        : <span id="ngramLength">8</span>
        <textarea id="taDensitySequence" style="width:100%;font-size:small;height:100">...</textarea>
        <table id="fcgrTable">
            <tr>
                <td id="fcgrForward">USM forward density ...</td>
                <td id="fcgrBackward">USM backward density...</td>
            </tr>
        </table>
    </div>
    <div id="divFold">
        <hr>
        <h2>C. USM fold language model</h2>
        <p>
        The sequence data provided above (B. Density Distribution) was encoded by the USM iterated map procedure into forward and backward numeric coordinates coordinates. This embedded space defines a language model that can now be used to inspect the statistics of unit emission given a new sequence.
        Two options are provided here, the new sequence can be provided directly as embedded coordinates, or an actual sequence can be used to produce those cordinates.
        </p>
        <p>
            <li><b>C1) Pick a position from the reference sequence </b><input type="number" id="picki" size="50px"><button id="pickn">pick from reference</button><input id="iRange" type="range"></li>
            <span id="pickedSeq" style="color:green">...</span>
            
            <li><b>C2) Embedding coordinates</b> (maybe start by picking from training data, above, and then try with your own sequences, below)</li>
            Forward - 1 = <span id="before" style="color:blue;"></span>( <input id="fw1"> , <input id="fw2"> ): bin [<input style="width: 4em;" id="binFw1" type='number'>,<input style="width: 4em;" id="binFw2" type='number'>] = <span id="countFw"></span>
            <br>Backward +1 = <span id="after" style="color:blue;"></span>( <input id="bk1"> , <input id="bk2"> ): bin [<input style="width: 4em;" id="binBk1" type='number'>,<input style="width: 4em;" id="binBk2" type='number'>] = <span id="countBk"></span>
            <p><h3>Estimate emission probablity using the language model</h3>
            <p>
                <table>
                    <tr>
                        <th>Nucleotide</th><th>Forward embedding</th><th>Backward embedding</th><th>Compound</th>
                    </tr>
                    <tr>
                        <th>A (0,0)</th><td><input id="fw1A" style="width:20em;"><br><input id="fw2A" style="width:20em;"><br>[<input style="width: 3em;" id="binAfw1" type='number'>,<input style="width: 3em;" id="binAfw2" type='number'>]</td><td><input id="bk1A" style="width:20em;"><br><input id="bk2A" style="width:20em;"><br>[<input style="width: 3em;" id="binAbk1" type='number'>,<input style="width: 3em;" id="binAbk2" type='number'>]</td><td>Compound</td>
                    </tr>
                    <tr>
                        <th>C (0,1)</th><td><input id="fw1C"><br><input id="fw2C"></td><td><input id="bk1C"><br><input id="bk2C"></td><td>Compound</td>
                    </tr>
                    <tr>
                        <th>G (1,0)</th><td><input id="fw1G"><br><input id="fw2G"></td><td><input id="bk1G"><br><input id="bk2G"></td><td>Compound</td>
                    </tr>
                    <tr>
                        <th>T (1,1)</th><td><input id="fw1T"><br><input id="fw2T"></td><td><input id="bk1T"><br><input id="bk2T"></td><td>Compound</td>
                    </tr>
                </table>
            </p>
            The estimation is made by determining the 4 possible embeded positions (the four nucleotides) that could be taken by the sequence being analysed, and counting how many reference sequence embedded positions were found in the corresponding 4 bins.
            </p>
            <li><b>C3) Using your own Sequence</b></li>
            <table>
                <tr>
                    <td>
                        Forward: <input size="30%">
                    </td>
                    <td>
                        <table>
                            <tr>
                                <td>r1c1</td><td>r1c2</td><td>r1c3</td>
                            </tr>
                        </table>
                    </td>
                    <td>
                        <input size="30%"> :Backward
                    </td>
                </tr>
            </table>
        </p>
    </div>
</div><script src="index.js"></script>
</body>
